src/GUI/UI/UIHouses.cpp

```diff
@@ -313,87 +313,102 @@ bool enterHouse(HOUSE_ID uHouseID) {
-        int am_pm_flag_open = 0;
-        int am_pm_flag_close = 0;
-        if (uOpenTime > 12) {
-            uOpenTime -= 12;
-            am_pm_flag_open = 1;
+    int openHours = buildingTable[uHouseID].uOpenTime;
+    int closeHours = buildingTable[uHouseID].uCloseTime;
+    GameTime currentTime = pParty->GetPlayingTime();
+    GameTime currentTimeDays = GameTime::FromDays(currentTime.GetDays());
+    bool isOpened = false;
+
+    if (closeHours > openHours) {
+        // Store opened within one day
+        GameTime openTime = currentTimeDays.AddHours(openHours);
+        GameTime closeTime = currentTimeDays.AddHours(closeHours);
+
+        isOpened = (currentTime >= openTime) && (currentTime <= closeTime);
+    } else {
+        // Store opens in one day and closes on next day
+        GameTime openTime = currentTimeDays.AddHours(openHours);
+        GameTime closeTime = currentTimeDays.AddHours(closeHours);
+
+        isOpened = (currentTime <= closeTime) || (currentTime >= openTime);
+    }
+
+    if (!isOpened) {
+        int amPmOpen = 0;
+        int amPmClose = 0;
+
+        if (openHours > 12) {
+            openHours -= 12;
+            amPmOpen = 1;
         }
-        if (uCloseTime > 12) {
-            uCloseTime -= 12;
-            am_pm_flag_close = 1;
+        if (closeHours > 12) {
+            closeHours -= 12;
+            amPmClose = 1;
         }
 
-        GameUI_SetStatusBar(LSTR_FMT_OPEN_TIME, uOpenTime, localization->GetAmPm(am_pm_flag_open), uCloseTime, localization->GetAmPm(am_pm_flag_close));
+        GameUI_SetStatusBar(LSTR_FMT_OPEN_TIME, openHours, localization->GetAmPm(amPmOpen), closeHours, localization->GetAmPm(amPmClose));
         if (pParty->hasActiveCharacter()) {
             pParty->activeCharacter().playReaction(SPEECH_STORE_CLOSED);
         }
 
         return false;
-    } else {
-        if (isShop(uHouseID)) {  // entering shops and guilds
-            if (!(pParty->PartyTimes.shopBanTimes[uHouseID]) || (pParty->PartyTimes.shopBanTimes[uHouseID] <= pParty->GetPlayingTime())) {
-                pParty->PartyTimes.shopBanTimes[uHouseID] = GameTime(0);
-            } else {
-                GameUI_SetStatusBar(LSTR_BANNED_FROM_SHOP);
-                return false;
-            }
+    }
+
+    if (isShop(uHouseID)) {
+        if (!(pParty->PartyTimes.shopBanTimes[uHouseID]) || (pParty->PartyTimes.shopBanTimes[uHouseID] <= pParty->GetPlayingTime())) {
+            pParty->PartyTimes.shopBanTimes[uHouseID] = GameTime(0);
+        } else {
+            GameUI_SetStatusBar(LSTR_BANNED_FROM_SHOP);
+            return false;
         }
+    }
 
+    uCurrentHouse_Animation = buildingTable[uHouseID].uAnimationID;
+    in_current_building_type = pAnimatedRooms[uCurrentHouse_Animation].uBuildingType;
+    if (in_current_building_type == BUILDING_THRONE_ROOM && pParty->uFine) {  // going to jail
+        uHouseID = HOUSE_JAIL;
         uCurrentHouse_Animation = buildingTable[uHouseID].uAnimationID;
```

Comment:
Can remove this line?

Reply:
uHouseID is changed here, so uCurrentHouse_Animation is setting again.